File: python/Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        number_of_zero = 0
        max_product = 1

        for num in nums:
            if num == 0:
                number_of_zero += 1
            else:
                max_product *= num

        results = []
        for num in nums:
            if number_of_zero > 1:
                results.append(0)
            elif number_of_zero == 1:
                if num == 0:
                    results.append(max_product)
                else:
                    results.append(0)
            else:
                results.append(max_product / num)
            
        return results

File: cpp/Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        int left = 1, right = 1;
        vector<int> result(n, 1);
        
        for (int i = 0; i < n; i++) {
            result[i] *= left;
            result[n - 1 - i] *= right;
            left *= nums[i];
            right *= nums[n - 1 -i];
        }
        
        return result;
    }
};
